clash-lib/src/Clash/Backend/SystemVerilog.hs

```diff
@@ -176,68 +166,31 @@ instance Backend SystemVerilogState where
-  ,"else","end","endcase","endchecker","endclass","endclocking","endconfig"
-  ,"endfunction","endgenerate","endgroup","endinterface","endmodule","endpackage"
-  ,"endprimitive","endprogram","endproperty","endspecify","endsequence"
-  ,"endtable","endtask","enum","event","eventually","expect","export","extends"
-  ,"extern","final","first_match","for","force","foreach","forever","fork"
-  ,"forkjoin","function","generate","genvar","global","highz0","highz1","if"
-  ,"iff","ifnone","ignore_bins","illegal_bins","implements","implies","import"
-  ,"incdir","include","initial","inout","input","inside","instance","int"
-  ,"integer","interconnect","interface","intersect","join","join_any"
-  ,"join_none","large","let","liblist","library","local","localparam","logic"
-  ,"longint","macromodule","matches","medium","modport","module","nand"
-  ,"negedge","nettype","new","nexttime","nmos","nor","noshowcancelled","not"
-  ,"notif0","notif1","null","or","output","package","packed","parameter","pmos"
-  ,"posedge","primitive","priority","program","property","protected","pull0"
-  ,"pull1","pulldown","pullup","pulsestyle_ondetect","pulsestyle_onevent"
-  ,"pure","rand","randc","randcase","randsequence","rcmos","real","realtime"
-  ,"ref","reg","reject_on","release","repeat","restrict","return","rnmos"
-  ,"rpmos","rtran","rtranif0","rtranif1","s_always","s_eventually","s_nexttime"
-  ,"s_until","s_until_with","scalared","sequence","shortint","shortreal"
-  ,"showcancelled","signed","small","soft","solve","specify","specparam"
-  ,"static","string","strong","strong0","strong1","struct","super","supply0"
-  ,"supply1","sync_accept_on","sync_reject_on","table","tagged","task","this"
-  ,"throughout","time","timeprecision","timeunit","tran","tranif0","tranif1"
-  ,"tri","tri0","tri1","triand","trior","trireg","type","typedef","union"
-  ,"unique","unique0","unsigned","until","until_with","untyped","use","uwire"
-  ,"var","vectored","virtual","void","wait","wait_order","wand","weak","weak0"
-  ,"weak1","while","wildcard","wire","with","within","wor","xnor","xor"]
-
-filterReserved :: Identifier -> Identifier
-filterReserved s = if s `elem` reservedWords
-  then s `TextS.append` "_r"
-  else s
-
 -- | Generate SystemVerilog for a Netlist component
-genVerilog :: Identifier -> SrcSpan -> HashMapS.HashMap Identifier Word -> Component -> SystemVerilogM ((String,Doc),[(String,Doc)])
-genVerilog _ sp seen c = preserveSeen $ do
-    Mon $ idSeen .= seen
+genSystemVerilog
+  :: ModName
+  -> SrcSpan
+  -> IdentifierSet
+  -> Component
+  -> SystemVerilogM ((String, Doc), [(String, Doc)])
+genSystemVerilog _ sp seen c = do
     -- Don't have type names conflict with module names
-    Mon $ tySeen %= HashMapS.unionWith max seen
+    Mon $ idSeen .= seen
+
+    -- Don't have type names conflict with type names generated in previous
+    -- genVHDL
+    typNames <- HashMap.elems <$> use nameCache
+    mapM_ (Id.addRaw . Id.toText) typNames
```

Comment:
I don't get this part... we're still in the same `SystemVerilogM` session, so why wouldn't previous calls to `genSystemVerilog` have already added the type names to the `identifierSet`?

Reply:
The goal of this PR was to move all logic concerning unique name generation to `Clash.Netlist.Id`. To that end, type names are generated using the `IdentifierSet` (stored at `idSeen` for `SystemVerilogState`). `nameCache` is just used to create a mapping between `HWType` and an `Identifier`. This has a side effect: when `idSeen` is reset duplicate names might be generated, so I re-add them here. Alternatively, `Mon $ idSeen .= seen` (just a couple of lines above the quote code) could do a union instead.

Reply:
Having said that.. I feel this is the "wrong" approach in the first place. Having this dependency between `genHDL` runs means we can't have parallel compilation of top entities (even after getting rid of the dependence of manifest files https://github.com/clash-lang/clash-compiler/pull/1227!). Instead, `IdentifierSet` should distinguish between global and local ids. Then, global ids should only be generated in their own namespace, e.g. prefixed by the module name. This would get rid of this^ quirk too.